TCRAnalysis/TCR_Analysis.R

```diff
@@ -0,0 +1,461 @@
+  }
+  
+  # error if not TCR spike-in
+  if (!grepl("TCR", dataset@experimentData@other[["PKCFileName"]], fixed = TRUE)) {
+    stop("Panel must include TCR spike-in to run this script")
+  }
+  
+  
+  #### QC using GeoMxTools
+  
+  # shift counts by one if not already done
+  if (dataset@experimentData@other[["shiftedByOne"]] == FALSE) {
+    dataset <- shiftCountsOne(dataset, useDALogic=TRUE)
+  }
+  
+  # need to do before running segment QC
+  pData(dataset)[, "Module"] <- dataset@experimentData@other[["PKCFileName"]]
+  
+  # run segment QC
+  dataset <- setSegmentQCFlags(dataset) 
+  
+ 
+  #### Filter segments and targets
+  
+  # filter segments
+  
+  if (filter_segments) {
+    lowNegatives <- which(pData(protocolData(dataset))[, "QCFlags"]$LowNegatives)
+    dataset = dataset[, -lowNegatives]
+  }
+  
+  # filter targets
+  
+  # get list of TCR relevant probes
+  tcr_probes <- fData(dataset)$TargetName[grepl('TR[A/B/D/G][C/J/V]', fData(dataset)$TargetName)]
+  
+  if (filter_targets) {
+    
+    expression_threshold <- ncol(exprs(dataset))*percent_thresh
+    low_targets <- which(rowSums(exprs(dataset) > 1) < expression_threshold)
+    targets_to_remove <- low_targets[!(low_targets %in% which(rownames(exprs(dataset)) %in% tcr_probes))]
+    
+    if (length(targets_to_remove) > 0) {
+      dataset = dataset[-targets_to_remove, ]
+    }
+  }
+  
+  
+  #### Normalize using GeoMxTools
+  
+  dataset <- normalize(dataset, norm_method="quant", 
+                               desiredQuantile = .9, toElt = "q_norm")
+  
+  
+  #### calculate detection of TCR probes over background (LOQ#)
+  # output logical matrix with every probe being noted as TRUE or FALSE 
+  # for being above (>) LOQ at the required thresholds specified by LOQ_thresholds
+
+  # calculate the negative geomean for each module
+  negativeGeoMeans <- 
+    esBy(negativeControlSubset(dataset), 
+         GROUP = "Module", 
+         FUN = function(x) { 
+           assayDataApply(x, MARGIN = 2, FUN = ngeoMean, elt = "exprs") 
+         }) 
+  pData(dataset)[, "NegGeoMean"] <- negativeGeoMeans
+  
+  # calculate the negative geo standard deviation for each module
+  negativeGeoSD <- 
+    esBy(negativeControlSubset(dataset), 
+         GROUP = "Module", 
+         FUN = function(x) { 
+           assayDataApply(x, MARGIN = 2, FUN = ngeoSD, elt = "exprs") 
+         }) 
+  pData(dataset)[, "NegGeoSD"] <- negativeGeoSD
+  
+  # initialize array to store results
+  LOQ_results <- array(NA, dim = c(nrow(exprs(dataset)), ncol(exprs(dataset)), length(LOQ_thresholds)), dimnames = list(rownames(exprs(dataset)), colnames(exprs(dataset)), paste0("LOQ", LOQ_thresholds)))
+  
+  # determine which TCR targets are above LOQ thresh and save to results array
+  for (i in 1:length(LOQ_thresholds)) {
+    
+    n = LOQ_thresholds[i]
+    LOQ <- data.frame(row.names = colnames(dataset))
+  
+    vars <- c("NegGeoMean", "NegGeoSD")
+    if(all(vars[1:2] %in% colnames(pData(dataset)))) {
```

Comment:
removed the if statement all together since geomean and sd are explicitly calculated at the beginning.  If there is an issue, it will error before this step